api/v1beta1/disruption_types.go

```diff
@@ -91,13 +94,85 @@ func init() {
 }
 
 // Hash returns the disruption spec JSON hash
-func (ds *DisruptionSpec) Hash() (string, error) {
+func (s *DisruptionSpec) Hash() (string, error) {
 	// serialize instance spec to JSON
-	specBytes, err := json.Marshal(ds)
+	specBytes, err := json.Marshal(s)
 	if err != nil {
 		return "", fmt.Errorf("error serializing instance spec: %w", err)
 	}
 
 	// compute bytes hash
 	return fmt.Sprintf("%x", md5.Sum(specBytes)), nil
 }
+
+func (s *DisruptionSpec) Validate() error {
+	// Rule: no targeted container if disruption is node-level
+	if len(s.Containers) > 0 && s.Level != chaostypes.DisruptionLevelPod {
+		return errors.New("cannot target specific container of a node-level disruption")
+	}
+
+	// Rule: at least one disruption field
+	if s.DNS == nil &&
+		s.CPUPressure == nil &&
+		s.Network == nil &&
+		s.NodeFailure == nil &&
+		s.DiskPressure == nil {
+		return errors.New("cannot apply a disruption with no target process")
+	}
+
+	//Rule: DNS and Network disruptions are incompatible
+	if s.DNS != nil && s.Network != nil {
+		return errors.New("cannot apply DNS and Network disruptions concurrently")
+	}
+
+	return nil
+}
+
+// validates rules for disruption global scope and all subsequent disruption specifications
+func (s *DisruptionSpec) ValidateDisruptionSpec() error {
```

Comment:
Having both `DisruptionSpec.Validate()` and `DisruptionSpec.ValidateDisruptionSpec()` is a bit confusing. I see that this method calls `Validate()`, I would suggest the highest level function takes the most general name, so this would be `Validate()` and the other would become ~`ValidateTopLevelOptions()` (or something else, it isnt like that's a good name). 

Also, remember that if a method shouldn't be called externally, it shouldn't be exported. I haven't finished reading, but if what is now `Validate()` is only called within `ValidateDisruptionSpec()`, it's best to lowercase the name, so that it isn't exported.

Reply:
Makes sense, both are now swapped and renamed (sort of)